SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialSpanIdStore.cpp

```diff
@@ -0,0 +1,204 @@
+				Update.FieldId = FieldId;
+				Update.NewSpanId = Op.span_id;
+				Update.OldSpanId = SpawnIdRef->SpanId;
+				FieldCollisions.Add(Update);
+			}
+		}
+
+		WriteSpanId(Id, FieldId, Op.span_id);
+	}
+
+	return FieldCollisions;
+}
+
+void SpatialSpanIdStore::WriteSpanId(const EntityComponentId& Id, const uint32 FieldId, Trace_SpanId SpanId)
+{
+	FieldIdMap& SpanIdMap = EntityComponentFieldSpanIds.FindOrAdd(Id);
+	EntityComponentFieldIdUpdateSpanId& SpawnIdRef = SpanIdMap.FindOrAdd(FieldId);
+	SpawnIdRef.SpanId = SpanId;
+	SpawnIdRef.UpdateTime = FDateTime::Now();
+}
+
+Trace_SpanId SpatialSpanIdStore::GetSpanId(const EntityComponentId& Id, const uint32 FieldId)
+{
+	Trace_SpanId ReturnSpanId = Trace_SpanId();
+
+	FieldIdMap* SpanIdMap = EntityComponentFieldSpanIds.Find(Id);
+	if (SpanIdMap == nullptr)
+	{
+		return ReturnSpanId;
+	}
+
+	EntityComponentFieldIdUpdateSpanId* UpdateSpanId = SpanIdMap->Find(FieldId);
+	if (UpdateSpanId == nullptr)
+	{
+		return ReturnSpanId;
+	}
+
+	return UpdateSpanId->SpanId;
+}
+
+Trace_SpanId SpatialSpanIdStore::GetMostRecentSpanId(const EntityComponentId& Id)
+{
+	FieldIdMap* SpanIdMap = EntityComponentFieldSpanIds.Find(Id);
+	if (SpanIdMap == nullptr)
+	{
+		return Trace_SpanId();
+	}
+
+	EntityComponentFieldIdUpdateSpanId MostRecent;
+	for (const auto& Pair : *SpanIdMap)
+	{
+		const EntityComponentFieldIdUpdateSpanId& Data = Pair.Value;
+		if (Data.UpdateTime > MostRecent.UpdateTime || Trace_SpanId_Equal(MostRecent.SpanId, Trace_SpanId()))
+		{
+			MostRecent = Data;
+		}
+	}
+
+	return MostRecent.SpanId;
+}
+
+bool SpatialSpanIdStore::DropSpanIds(const EntityComponentId& Id)
+{
+	return EntityComponentFieldSpanIds.Remove(Id) > 0;
+}
+
+bool SpatialSpanIdStore::DropSpanId(const EntityComponentId& Id, const uint32 FieldId)
+{
+	return DropSpanIdInternal(EntityComponentFieldSpanIds.Find(Id), Id, FieldId);
+}
+
+bool SpatialSpanIdStore::DropSpanIdInternal(FieldIdMap* SpanIdMap, const EntityComponentId& Id, const uint32 FieldId)
+{
+	if (SpanIdMap == nullptr)
+	{
+		return false;
+	}
+
+	bool bDropped = SpanIdMap->Remove(FieldId) > 0;
+	if (bDropped && SpanIdMap->Num() == 0)
+	{
+		EntityComponentFieldSpanIds.Remove(Id);
+	}
+
+	return bDropped;
+}
+
+void SpatialSpanIdStore::DropOldSpanIds()
+{
+	if (NextClearTime < FDateTime::Now())
+	{
+		UpdateNextClearTime();
+
+		FDateTime DropDateTime = FDateTime::Now() - FTimespan::FromSeconds(MinSpanIdLifetime);
+
+		TArray<EntityComponentFieldId> EntityComponentFieldIdsToDrop;
+		int32 NumDropped = 0;
+		bool bShouldBreak = false;
+
+		for (const auto& Pair : EntityComponentFieldSpanIds)
+		{
+			const EntityComponentId& Id = Pair.Key;
+			const FieldIdMap& Map = Pair.Value;
+
+			for (const auto& UpdateSpanIdPair : Map)
+			{
+				uint32 FieldId = UpdateSpanIdPair.Key;
+				const EntityComponentFieldIdUpdateSpanId& UpdateSpanId = UpdateSpanIdPair.Value;
+
+				if (UpdateSpanId.UpdateTime < DropDateTime)
+				{
+					EntityComponentFieldIdsToDrop.Add({ Id, FieldId });
```

Comment:
Think you can use a TMap iterator here and It.RemoveCurrent() to avoid storing them and deleting later